### PR TITLE
Add Dockerfile for building 0.5.0

### PR DESCRIPTION
**Pull request by johanbrandhorst**

### CODE REVIEW

**Comment by johanbrandhorst** [comment]
The latest image has been tested with the smoke test builds:
```
root@828c1680f342:/# tinygo build -size short -o test.elf -target=arduino             examples/blinky1
   code    data     bss |   flash     ram
    792      54     658 |     846     712
root@828c1680f342:/# tinygo build -size short -o test.elf -target=digispark           examples/blinky1
   code    data     bss |   flash     ram
    342       0     132 |     342     132
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10040            examples/blinky1
   code    data     bss |   flash     ram
   1688       4    2196 |    1692    2200
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10040            examples/blinky2
   code    data     bss |   flash     ram
   3348       4    2244 |    3352    2248
root@828c1680f342:/# tinygo build             -o blinky2                              examples/blinky2
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10040            examples/test
   code    data     bss |   flash     ram
   6368     128    2196 |    6496    2324
root@828c1680f342:/# tinygo build -size short -o test.elf -target=microbit            examples/echo
   code    data     bss |   flash     ram
   2712       0    2228 |    2712    2228
root@828c1680f342:/# tinygo build -size short -o test.elf -target=nrf52840-mdk        examples/blinky1
   code    data     bss |   flash     ram
   1416       4    4244 |    1420    4248
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10031            examples/blinky1
   code    data     bss |   flash     ram
   1128       0    2196 |    1128    2196
root@828c1680f342:/# tinygo build -size short -o test.elf -target=bluepill            examples/blinky1
   code    data     bss |   flash     ram
   2392       0    2228 |    2392    2228
root@828c1680f342:/# tinygo build -size short -o test.elf -target=reelboard           examples/blinky1
   code    data     bss |   flash     ram
   1416       4    4244 |    1420    4248
root@828c1680f342:/# tinygo build -size short -o test.elf -target=reelboard           examples/blinky2
   code    data     bss |   flash     ram
   3092       4    4292 |    3096    4296
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10056            examples/blinky1
   code    data     bss |   flash     ram
   1416       4    4244 |    1420    4248
root@828c1680f342:/# tinygo build -size short -o test.elf -target=pca10056            examples/blinky2
   code    data     bss |   flash     ram
   3092       4    4292 |    3096    4296
root@828c1680f342:/# tinygo build -size short -o test.elf -target=itsybitsy-m0        examples/blinky1
   code    data     bss |   flash     ram
  15752      32    4452 |   15784    4484
root@828c1680f342:/# tinygo build -size short -o test.elf -target=circuitplay-express examples/blinky1
   code    data     bss |   flash     ram
  15752      32    4452 |   15784    4484
root@828c1680f342:/# tinygo build -size short -o test.elf -target=stm32f4disco        examples/blinky1
   code    data     bss |   flash     ram
   2728       0    4260 |    2728    4260
root@828c1680f342:/# tinygo build -size short -o test.elf -target=stm32f4disco        examples/blinky2
   code    data     bss |   flash     ram
   4276       0    4276 |    4276    4276
root@828c1680f342:/# tinygo build             -o wasm.wasm -target=wasm               examples/wasm/export
root@828c1680f342:/# tinygo build             -o wasm.wasm -target=wasm               examples/wasm/main
```

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ENV PATH=${PATH}:/usr/local/go/bin
+
+ENV TINYGO_RELEASE=0.5.0
+RUN wget https://github.com/tinygo-org/tinygo/releases/download/v${TINYGO_RELEASE}/tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz && \
+    tar xfv tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz -C /usr/local && \
+    rm tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz
+ENV PATH=${PATH}:/usr/local/tinygo/bin
+
+RUN apt-get remove -y wget gnupg && \
+    apt-get autoremove -y
```
Comment (aykevl) [inline Dockerfile]: You can try `apt-get clean` to clean the cache, it might bring down the size a bit.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
+FROM debian:stable-slim
+
+RUN apt-get update && \
+    apt-get install -y wget gcc gcc-avr avr-libc gnupg
+
+RUN echo 'deb http://apt.llvm.org/stretch/ llvm-toolchain-stretch-8 main' > /etc/apt/sources.list.d/clang-8.list
+RUN wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+RUN apt-get update && apt-get install -y clang-8
```
Comment (aykevl) [inline Dockerfile]: You could try `--no-install-recommends` to only install real dependencies.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ENV GO_RELEASE=1.12.4
+RUN wget https://dl.google.com/go/go${GO_RELEASE}.linux-amd64.tar.gz && \
+    tar xfv go${GO_RELEASE}.linux-amd64.tar.gz -C /usr/local && \
+    rm go${GO_RELEASE}.linux-amd64.tar.gz
+ENV PATH=${PATH}:/usr/local/go/bin
+
+ENV TINYGO_RELEASE=0.5.0
+RUN wget https://github.com/tinygo-org/tinygo/releases/download/v${TINYGO_RELEASE}/tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz && \
+    tar xfv tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz -C /usr/local && \
+    rm tinygo${TINYGO_RELEASE}.linux-amd64.tar.gz
```
Comment (aykevl) [inline Dockerfile]: Only /usr/local/go/src is required, the rest may be removed.

File: Dockerfile (hunk 2 above)
Comment (johanbrandhorst) [reply Dockerfile]: Fixed

File: Dockerfile (hunk 3 above)
Comment (johanbrandhorst) [reply Dockerfile]: Fixed

**Comment by johanbrandhorst** [comment]
Latest image confirmed working with smoke test

**Comment by johanbrandhorst** [comment]
We're at 886MB now:
```
$ docker images
REPOSITORY                TAG                 IMAGE ID            CREATED              SIZE
<none>                    <none>              001c1cee0d41        About a minute ago   886MB
```

**Comment by deadprogram** [comment]
This is great. Also interesting how little is really needed, by which I mean it looks like we do not need some of the dependencies listed in https://tinygo.org/getting-started/linux/#arm-cortex-m

Merging.

**Comment by aykevl** [comment]
> Also interesting how little is really needed, by which I mean it looks like we do not need some of the dependencies listed in https://tinygo.org/getting-started/linux/#arm-cortex-m

Yes, the other 3 dependencies (llvm-8-dev lld-8 libclang-8-dev) are only necessary when building from source and want to link to these system-wide LLVM libraries. I missed that during the docs review.

**Pull request merged**